feat(ui): add tooltips for vitals percents

alxu-sentry opened the pull request:
Not super happy about the code structure, open to suggestions. Seems to work fine though.

alxu-sentry committed:
feat(ui): add tooltips for vitals percents

Zylphrex commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -26,13 +29,46 @@ export default function VitalPercents(props: Props) {
+                title={
+                  Array.isArray(props.vital)
+                    ? 'Good'
+                    : tct('Good: <[threshold]ms', {
```
This tooltip doesn't explain much more than what the visuals already show. Can be elaborate a little more on what this means? For example (and feel free to change it up)

> 40% of first contentful paint is below 2400ms.

alxu-sentry replied:
I thought about that, but I'm not sure how much additional information it adds. The percentage is already displayed right next to the tooltip, and the name of the metric is in large print on the left side of the table. The main point of this tooltip IMO is to add the threshold information. Maybe that should just be rendered directly and not hidden in a tooltip though?

alxu-sentry replied:
had a chat with Adam, we agreed that always showing it adds unnecessary noise to the page. i'm not that happy with it either, but i think it's fine and not that important, so let's merge it on Tuesday if nobody has any better ideas over the weekend?

Zylphrex commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -26,13 +29,46 @@ export default function VitalPercents(props: Props) {
+              <Tooltip
+                title={
+                  Array.isArray(props.vital)
+                    ? 'Good'
```
Just saying "Good" here doesn't add any additional value. Don't have a better suggestion at the moment but will update if I think of anything.

billyvg replied:
We should also wrap these strings with `t()`

alxu-sentry replied:
I had that impression too about "Good". I mainly wanted to show the standard words (maybe someone sees this and googles "web vitals good needs improvement poor"?), but I agree that it's not terribly useful. Agreed on `t`.

markstory commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -26,13 +29,46 @@ export default function VitalPercents(props: Props) {
+              <Tooltip
+                title={
+                  Array.isArray(props.vital)
+                    ? 'Needs improvement'
```
```suggestion
                    ? t('Needs improvement')
```
Display text should be in t() so we could translate it in the future.

markstory commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -26,13 +29,46 @@ export default function VitalPercents(props: Props) {
+              <Tooltip
+                title={
+                  Array.isArray(props.vital)
+                    ? 'Poor'
```
```suggestion
                    ? t('Poor')
```

alxu-sentry committed:
fix icon vertical alignment

alxu-sentry committed:
move Tooltip outside, refactor everything

putting the Tooltip inside was causing the icons to be misaligned.
vertical-align: middle fixes it, but @Zylphrex suggested moving the Tooltip
outside. I think this is a better solution anyways, so implemented it. it made
the code too duplicated though, so I also refactored it.

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -38,7 +38,10 @@ export default function VitalPercents(props: Props) {
               >
-                <IconFire color={vitalStateColors[p.vitalState] as Color} />
+                <IconFire
+                  style={{verticalAlign: 'middle'}}
```
Instead of inline styles you can use `css` from `@emotion/core` and apply it to each of these Icons using the `css` prop.

alxu-sentry replied:
I totally rewrote this in v3 of this PR. Could you possibly review it again? Thanks!

billyvg replied:
Oh, was I looking at an old commit or something? My bad

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -38,7 +38,10 @@ export default function VitalPercents(props: Props) {
-                <IconFire color={vitalStateColors[p.vitalState] as Color} />
+                <IconFire
+                  style={{verticalAlign: 'middle'}}
+                  color={vitalStateColors[p.vitalState] as Color}
```
Should this be cast as `Color` where we export `vitalStateColors`?

alxu-sentry replied:
Possibly. I'll look into it.

alxu-sentry replied:
We can't just change `vitalStateColors` to be `Color`s, because it's also used at:

https://github.com/getsentry/sentry/blob/f2e173d1a93b1da03b819c3cb02e437a4575c8ca/src/sentry/static/sentry/app/views/performance/vitalDetail/table.tsx#L430

I think this is fine for now, and if we want to change vitalStateColors that can be done in a separate PR.

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -1,44 +1,67 @@
   return (
     <VitalSet>
       {props.percents.map(p => {
+        const VitalStateIcon = vitalStateIcons[p.vitalState];
```
`p` is a confusing var name in this case - we use `p` a lot in styled components to prefer to `props`. `percent` would make more sense here (or even destructuring it so that you don't need to deal with an intermediate name).

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -1,44 +1,67 @@
 import space from 'app/styles/space';
+import {WebVital} from 'app/utils/discover/fields';
 import {formatPercentage} from 'app/utils/formatters';
 import theme, {Color} from 'app/utils/theme';
```
We should not import `theme` like this, instead it is passed in through `props` when you are creating a styled component.

alxu-sentry replied:
Yes, I ran into this issue with https://github.com/getsentry/sentry/pull/23893. I think it's fine for now though; if we want to change that I think it should be in a separate PR.

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -1,44 +1,67 @@
+  [VitalState.GOOD]: IconCheckmark,
+};
+
+function vitalStateText(vital, vitalState) {
```
Personal pref here but I thinking adding a verb here would give a bit more context of what it's doing, e.g. `getVitalStateText`

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/vitalPercents.tsx:
```diff
@@ -1,44 +1,67 @@
+            <VitalStatus>
+              <VitalStateIcon color={vitalStateColors[p.vitalState] as Color} />
+              <span>
+                {props.showVitalPercentNames && t(`${p.vitalState}`)}{' '}
```
I saw the thread in slack, but wanted to address it here for posterity: using `t()` with variables will function fine, but the translation part will do nothing.

alxu-sentry replied:
I'm virtually certain this is dead code. I have another PR lined up to delete showVitalPercentNames entirely, but it obviously merge-conflicts with this one.

Zylphrex replied:
This is used [here](https://github.com/getsentry/sentry/blob/master/src/sentry/static/sentry/app/views/performance/vitalDetail/vitalDetailContent.tsx#L183-L188), thought an argument can be made that with the tooltips, it's okay to remove this anyways.

alxu-sentry replied:
Ah, I see. I think it should be consistent everywhere anyways, and the tooltip is probably clear enough that we can delete it entirely.

billyvg reviewed (commented):
I left some comments that are unrelated to your PR (so no need to fix them), but thought it'd be helpful info for you. Otherwise, LGTM

alxu-sentry committed:
re-add checkmark circle, use utils vitalStateIcons

alxu-sentry committed:
refactor to use type-safe Colors

alxu-sentry committed:
resolve remaining nits

- remove last use of theme
- vitalStateText -> getVitalStateText
- p -> pct

alxu-sentry committed:
fix copy/paste oopsies

alxu-sentry committed:
Merge remote-tracking branch 'origin/master' into vitals-percent-tooltip

alxu-sentry committed:
fix vital bar regression

alxu-sentry committed:
Merge remote-tracking branch 'origin/master' into vitals-percent-tooltip

alxu-sentry committed:
actually fix bar color

billyvg commented on src/sentry/static/sentry/app/views/performance/vitalDetail/utils.tsx:
```diff
@@ -33,20 +33,18 @@ export enum VitalState {
+  [VitalState.POOR]: <IconFire color={vitalStateColors[VitalState.GOOD]} />,
+  [VitalState.MEH]: <IconWarning color={vitalStateColors[VitalState.GOOD]} />,
+  [VitalState.GOOD]: (
+    <IconCheckmark color={vitalStateColors[VitalState.GOOD]} isCircled />
```
Should these all be the same color?

alxu-sentry replied:
absolutely not, but i fixed that already

alxu-sentry replied:
could you review it, uh... again again? :/

k-fish reviewed (approved):
Looks good Alex 👏
